pandas/core/arrays/text.py

```diff
@@ -0,0 +1,279 @@
+        value = extract_array(value, extract_numpy=True)
+        if isinstance(value, type(self)):
+            # extract_array doesn't extract PandasArray subclasses
+            value = value._ndarray
+
+        scalar_key = lib.is_scalar(key)
+        scalar_value = lib.is_scalar(value)
+        if scalar_key and not scalar_value:
+            raise ValueError("setting an array element with a sequence.")
+
+        # validate new items
+        if scalar_value:
+            if scalar_value is None:
+                value = np.nan
+            elif not (isinstance(value, str) or np.isnan(value)):
+                raise ValueError(
+                    "Cannot set value '{}' into a TextArray.".format(value)
+                )
+        else:
+            if not is_array_like(value):
+                value = np.asarray(value, dtype=object)
+            if len(value) and not lib.is_string_array(value, skipna=True):
+                raise ValueError("Must provide strings.")
+
+        super().__setitem__(key, value)
+
+    def fillna(self, value=None, method=None, limit=None):
+        # TODO: validate dtype
+        return super().fillna(value, method, limit)
+
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+        if isinstance(dtype, TextDtype):
+            if copy:
+                return self.copy()
+            return self
+        return super().astype(dtype, copy)
+
+    def _reduce(self, name, skipna=True, **kwargs):
+        if name == "sum":
+            vals = self._ndarray
+            missing = self.isna()
+            if skipna:
+                vals = vals[~missing]
+            elif missing.any():
+                return np.nan
+            return vals.sum()
+
+        raise TypeError("Cannot perform reduction '{}' with string dtype".format(name))
+
+    def value_counts(self, dropna=False):
+        from pandas import value_counts
+
+        return value_counts(self._ndarray, dropna=dropna)
+
+    # Overrride parent, because we have different return types.
+    @classmethod
+    def _create_arithmetic_method(cls, op):
+        def method(self, other):
+            if isinstance(other, (ABCIndexClass, ABCSeries, ABCDataFrame)):
+                return NotImplemented
+
+            elif isinstance(other, cls):
+                other = other._ndarray
+
+            mask = isna(self) | isna(other)
+            valid = ~mask
+
+            if not lib.is_scalar(other):
+                other = np.asarray(other)
+                other = other[valid]
+
+            result = np.empty_like(self._ndarray, dtype="object")
+            result[mask] = np.nan
+            result[valid] = op(self._ndarray[valid], other)
+
+            if op.__name__ in {"add", "radd", "mul", "rmul"}:
+                new = TextArray
+            elif mask.any():
+                new = lambda x: np.asarray(x, dtype="object")
+            else:
+                new = lambda x: np.asarray(x, dtype="bool")
```

Comment:
is there a viable way to do this without the lambdas?  i think if you added `dtype` to the constructor (which would probably better match the pattern with other EAs) then `new` can just be `np.asarray` and also define `dtype` here and below 257, 259